chore(gce): pass auth flags to KCM and KS

knight42 opened the pull request:
**What type of PR is this?**

/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:

Pass flags `--authentication-kubeconfig` and `--authorization-kubeconfig` to controller-manager and scheduler, so
that we could grab metrics from their secure ports in tests, otherwise we will get 403 response even if we access with the correct credential.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref #91506 #96216

**Special notes for your reviewer**:

Now #96216 is being blocked by the job `pull-kubernetes-e2e-gce-100-performance`, because in this job the Prometheus installed by [clusterloader2](https://github.com/kubernetes/perf-tests/tree/master/clusterloader2) would only grab metrics from the insecure port of controller-manager and scheduler, so this job would not succeed.

We now have to refactor `clusterloader2` to make it grab metrics from the secure port, but that would require controller-manager and scheduler have correctly set the flags `--authorization-kubeconfig` and `--authentication-kubeconfig` in the test, and that is done by this PR.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

knight42 commented:
/cc @liggitt 
/priority important-soon
/triage accepted

knight42 commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

liggitt commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -2126,7 +2127,8 @@ function start-kube-scheduler {
     create-kube-scheduler-config
```
what happens in this branch?

knight42 replied:
I am not sure, looks like the file `KUBE_SCHEDULER_CONFIG` is managed by some other users. In addition, `${KUBE_SCHEDULER_CONFIG}` is not set currently: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/96550/pull-kubernetes-e2e-gce-100-performance/1327123753032749056/artifacts/e2e-96550-95a39-master/kube-scheduler.log (`--config=""`)

liggitt replied:
not in CI, but by some launchers, it can be

@ahg-g, how does kube-scheduler configure these two flags when running from a config file?

ahg-g replied:
nothing different: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/app/options/options.go#L216

ahg-g replied:
shouldn't we pass those flags in both cases, when component-config is passed and when not?

knight42 replied:
given that the content of `KUBE_SCHEDULER_CONFIG` is managed by other users and I guess it is not desired to append these two flags in this script, I think it is okay to left as it is and let the user set these flags in the `KUBE_SCHEDULER_CONFIG`.

liggitt replied:
> and let the user set these flags in the `KUBE_SCHEDULER_CONFIG`.

can those flags be set via that config file?

knight42 replied:
> can those flags be set via that config file?

hmmm after looking into the source code, I don't think so..:
https://github.com/kubernetes/kubernetes/blob/0913f8a54cc68164d5182796476a9f608321ac17/pkg/scheduler/apis/config/types.go#L55-L112
I should take a look first..

And regarding the config file, user might decide to set the path to kubeconfig by setting `.clientConnection.kubeconfig` in the config file, and it is not quite easy to extract that path from `KUBE_SCHEDULER_CONFIG`.

knight42 replied:
As now `--kubeconfig` could be set via the config file, is it reasonable to set `--authentication-kubeconfig` and `--authorization-kubeconfig` via the config file?
cc @ahg-g

ahg-g replied:
Adding those flags to the config is probably a question to component config working group, those seem to be generic enough to be placed in a common type embedded in all configs.

knight42 replied:
@liggitt now that the v1.20 restriction has been lifted, I want to pick up the work now, is it required to add those flags to kube-scheduler config?

liggitt replied:
It has to work whether we start from a config file or from flags. I'll defer to the scheduler team on how to accomplish that.

knight42 replied:
@ahg-g @liggitt Now this should work for both cases, PTAL

liggitt commented:
one question, looks reasonable otherwise.


/milestone v1.21

liggitt commented:
/assign @ahg-g 
/sig scheduler

knight42 commented:
@liggitt Are there any chance to get this landed in v1.20? This is more likely to be a test-only change, and with this change I could start to refactor `clusterloader2` to make it grab metrics from the secure port.

liggitt commented:
> Are there any chance to get this landed in v1.20? This is more likely to be a test-only change, and with this change I could start to refactor `clusterloader2` to make it grab metrics from the secure port.

No, it's too late in the cycle to be making optional deployment changes like this

knight42 committed:
chore(gce): pass auth flags to KCM and KS

Pass flags `--authentication-kubeconfig` and
`--authorization-kubeconfig` to controller-manager and scheduler,
so that we could grab metrics from their secure ports in tests.

knight42 commented:
/retest

liggitt commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -2111,9 +2112,15 @@ function start-kube-scheduler {
   if [[ -n "${KUBE_SCHEDULER_CONFIG:-}" ]]; then
     create-kube-scheduler-config
+    local config_path
+    config_path=$(awk '/kubeconfig:/{print $2;}' /etc/srv/kubernetes/kube-scheduler/config)
```
rather than pulling out the kubeconfig file, should we just use the one created above in `create-kubeconfig "kube-scheduler" "${KUBE_SCHEDULER_TOKEN}"` and move the authentication-kubeconfig/authorization-kubeconfig flags outside of the KUBE_SCHEDULER_CONFIG if/else blocks?

knight42 replied:
wow thanks for pointing it out 👍 , I should not have missed that. I have updated the script as you said, PTAL.

knight42 commented:
/retest

knight42 commented:
/retest

liggitt commented:
/lgtm
/approve
